### PR TITLE
Fix Race condition for SQL Connection

### PR DESCRIPTION
**Pull request by Psycast**
Song Options context menu doesn't always show up when first coming from the login/game loading screen because it's connection is async. But after a song result, the it's already connected so it appears normally.
Fix is just adding a event listener to the connection and toggling the visibility of the menu item instead of add/removing it.

### CODE REVIEW

File: src/GlobalVariables.as [hunk 1]
```diff
@@ -158,22 +160,24 @@ package
...
-                sql_connect = false;
-                trace("Error message:", event.error.message);
-                trace("Details:", event.error.details);
-            }
+        private function sql_errorHandler(event:SQLErrorEvent):void
+        {
+            sql_connect = false;
+            trace("Error message:", event.error.message);
+            trace("Details:", event.error.details);
+            this.dispatchEvent(new Event(DB_CONNECT_ERROR));
```
Comment (G-flat) [inline src/GlobalVariables.as:180]: Being a bit nitpicky here, but I would remove the event listeners when either of these functions are called as follows:

```
private function sql_openHandler(event:SQLEvent):void
{
    sql_conn.removeEventListener(SQLEvent.OPEN, sql_openHandler);
    sql_conn.removeEventListener(SQLErrorEvent.ERROR, sql_errorHandler);

    sql_connect = true;

    trace("Database was connected successfully");

    SQLQueries.refreshStatements(sql_conn);
    this.dispatchEvent(new Event(DB_CONNECT_COMPLETE));
}

private function sql_errorHandler(event:SQLErrorEvent):void
{
    sql_conn.removeEventListener(SQLEvent.OPEN, sql_openHandler);
    sql_conn.removeEventListener(SQLErrorEvent.ERROR, sql_errorHandler);

    sql_connect = false;
    trace("Error message:", event.error.message);
    trace("Details:", event.error.details);
    this.dispatchEvent(new Event(DB_CONNECT_ERROR));
}
```

File: src/menu/MenuSongSelection.as [hunk 2]
```diff
@@ -112,6 +113,9 @@ package menu
 
         override public function init():Boolean
         {
+            // Listen for DB Connection
+            _gvars.addEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
+
```
Comment (G-flat) [inline src/menu/MenuSongSelection.as:118]:
```suggestion
            // Listen for DB Connection
            _gvars.addEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
            _gvars.addEventListener(GlobalVariables.DB_CONNECT_ERROR, e_sqlConnectError);
```

File: src/menu/MenuSongSelection.as [hunk 3]
```diff
@@ -340,6 +342,8 @@ package menu
 
         override public function stageRemove():void
         {
+            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
+
```
Comment (G-flat) [inline src/menu/MenuSongSelection.as:346]:
```suggestion
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_ERROR, e_sqlConnectError);
```

File: src/menu/MenuSongSelection.as [hunk 4]
```diff
@@ -1719,6 +1723,16 @@ package menu
...
+        /**
+         * Called when the SQL Local DB is connected.
+         * @param event
+         */
+        private function e_sqlConnectComplete(event:Object):void
+        {
+            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
+            songItemSongOptionsContext.visible = _gvars.sql_connect; // Race Condition
+        }
+
```
Comment (G-flat) [inline src/menu/MenuSongSelection.as:1735]:
```suggestion
        /**
         * Called when the SQL Local DB is connected.
         * @param event
         */
        private function e_sqlConnectComplete(event:Object):void
        {
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_ERROR, e_sqlConnectError);
            songItemSongOptionsContext.visible = _gvars.sql_connect; // Race Condition
        }

        /**
         * Called when the SQL Local DB cannot be connected.
         * @param event
         */
        private function e_sqlConnectError(event:Object):void
        {
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_COMPLETE, e_sqlConnectComplete);
            _gvars.removeEventListener(GlobalVariables.DB_CONNECT_ERROR, e_sqlConnectError);
            songItemSongOptionsContext.visible = _gvars.sql_connect; // Race Condition
        }
```

File: src/menu/MenuSongSelection.as (hunk 2 above)
Comment (Psycast) [reply src/menu/MenuSongSelection.as:118]: We actually don't need to listen for a DB connection error here, as we only care if it connects successfully.

File: src/GlobalVariables.as [hunk 5]
```diff
@@ -155,25 +162,42 @@ package
...
+
+            this.dispatchEvent(new Event(DB_CONNECT_ERROR));
+        }
+
+        private function sql_errorHandler(event:SQLErrorEvent):void
+        {
+            trace("SQL Execution Error:");
+            trace("Error message:", event.error.message);
+            trace("Details:", event.error.details);
         }
```
Comment (G-flat) [inline src/GlobalVariables.as:201]:
```suggestion
        private function sql_errorHandler(event:SQLErrorEvent):void
        {
            sql_conn.removeEventListener(SQLErrorEvent.ERROR, sql_errorHandler);
            trace("SQL Execution Error:");
            trace("Error message:", event.error.message);
            trace("Details:", event.error.details);
        }
```

**Review by G-flat — APPROVED** [review]

**Pull request merged**